lldb/tools/lldb-dap/DAPLog.h

```diff
@@ -9,44 +9,66 @@
       ::std::chrono::duration<double> now{                                     \
           ::std::chrono::system_clock::now().time_since_epoch()};              \
-      *log_private << ::llvm::formatv("{0:f9} ", now.count()).str()            \
-                   << ::llvm::formatv(__VA_ARGS__).str() << std::endl;         \
+      ::std::string out;                                                       \
+      ::llvm::raw_string_ostream os(out);                                      \
+      os << ::llvm::formatv("{0:f9} ", now.count()).str()                      \
+         << ::llvm::formatv(__VA_ARGS__).str() << "\n";                        \
+      log_private->WriteMessage(out);                                          \
     }                                                                          \
   } while (0)
 
 // Write message to log, if error is set. In the log message refer to the error
 // with {0}. Error is cleared regardless of whether logging is enabled.
 #define DAP_LOG_ERROR(log, error, ...)                                         \
   do {                                                                         \
-    ::std::ofstream *log_private = (log);                                      \
+    ::lldb_dap::Log *log_private = (log);                                      \
     ::llvm::Error error_private = (error);                                     \
     if (log_private && error_private) {                                        \
       ::std::chrono::duration<double> now{                                     \
           std::chrono::system_clock::now().time_since_epoch()};                \
-      *log_private << ::llvm::formatv("{0:f9} ", now.count()).str()            \
-                   << ::lldb_dap::FormatError(::std::move(error_private),      \
-                                              __VA_ARGS__)                     \
-                   << std::endl;                                               \
+      ::std::string out;                                                       \
+      ::llvm::raw_string_ostream os(out);                                      \
+      os << ::llvm::formatv("{0:f9} ", now.count()).str()                      \
+         << ::lldb_dap::FormatError(::std::move(error_private), __VA_ARGS__)   \
+         << "\n";                                                              \
+      log_private->WriteMessage(out);                                          \
     } else                                                                     \
       ::llvm::consumeError(::std::move(error_private));                        \
   } while (0)
 
 namespace lldb_dap {
 
+class Log {
+public:
+  Log(std::ofstream stream) : m_stream(std::move(stream)) {}
+
+  void WriteMessage(llvm::StringRef message) {
+    std::scoped_lock<std::mutex> lock(m_mutex);
+    m_stream << message.str();
+    m_stream.flush();
+  }
+
+private:
+  std::mutex m_mutex;
+  std::ofstream m_stream;
```

Comment:
I hate to give you more work, but I think it's probably time we switch over to `llvm::raw_ostream`. That way we can avoid the string conversion in `WriteMessage` above, though for this PR this can be avoided by moving the buffer into the write function as I suggested earlier.

Reply:
Moved from `std::ofstream` to `llvm::raw_fd_ostream`